WIP: Update base to U-Boot 2023.07

samueldr opened the pull request:
Feature branches here:

 - https://github.com/Tow-Boot/U-Boot/branches/all?query=tow-boot%2F2023.07%2F

### TODO

We'll need to tag the current state of the Tow-Boot fork, whether or not it is what is released as `2023.07`.

 - [x] Tag `2023.07-007-rc1` in the U-Boot repo with the accepted state
 - [x] Change impl. to work with that tag while in rc


### Known issues

<details><summary>"Resolved" rockchip issue</summary>

#### Rockchip

Rockchip moved to "standard[sic] boot" breaking the assumptions of the boot menu builder.

AFAIK this is the only actual real-world target/family that has moved to it.

A solution will need to be found, I haven't spent much time thinking about it yet.

The first thought I had is that *since anyway* we have to mess with distroboot things for most targets, we might benefit from making a "unified" distroboot env within our fork, instead of relying on the board-specific distroboot builders. And on top of that, making the *order* they are tried a `CONFIG_TOW_BOOT_DISTROBOOT_TARGETS` option, meaning that the boot order and targets can be handled through direct configuration rather than messy patches and `#ifdef`s.

That solution would be a stop-gap until the new boot methods are more universally enabled (maybe we can help down the line) and we would instead use the data from bootstd directly without going through cli commands when building the UI.

</details>

* * *

### Changelog

 - "Resolved" the Rockchip conundrum https://github.com/Tow-Boot/U-Boot/compare/d705eb1a60dbc6d507974afbf1c26663c321077c...61f7ac4831c93f74f2f542966d072f4aa62fc5cc

samueldr committed:
tow-boot/builder: Add pyelftool

samueldr committed:
tow-boot/options: Fix VIDEO_LOGO_MAX_SIZE

The Kconfig infra will provide a bogus value instead of the correct
value when it is not already in `0x...` hex format.

*sigh*

samueldr committed:
tow-boot/config: Use HIDE_LOGO_VERSION

samueldr committed:
tow-boot/src: Bump to 2023.07

shvetsnikita commented:
From my quick testing on a couple of already supported boards (Renegade Elite and OrangePi PC) I noticed that when Tow-Boot ask to press Esc or Ctrl+C, keyboard doesn't work at all (even no reaction to NumLock, so not powered?) but in GRUB2 keyboard works fine.

shvetsnikita commented:
Another observation is that SPI install doesn't work on Renegade Elite.

shvetsnikita commented:
At least on Renegade Elite Linux does not detect USB controllers (lsusb is empty) unless I use kernel's own dtb. Relevant messages in dmesg:

```
[    5.216389] usbcore: registered new interface driver usbfs
[    5.216510] usbcore: registered new interface driver hub
[    5.216625] usbcore: registered new device driver usb
[    5.689858] OF: graph: no port node found in /i2c@ff160000/usb-typec@22
[    5.897673] OF: graph: no port node found in /i2c@ff3d0000/usb-typec@22
[   16.106883] platform fe800000.usb: deferred probe pending
[   16.106894] platform fe900000.usb: deferred probe pending
[   16.106905] platform fe3a0000.usb: deferred probe pending
[   16.106913] platform fe3e0000.usb: deferred probe pending
[   16.106922] platform fe380000.usb: deferred probe pending
[   16.106930] platform fe3c0000.usb: deferred probe pending
[   16.106939] platform ff770000.syscon:usb2phy@e450: deferred probe pending
[   16.106947] platform ff770000.syscon:usb2phy@e460: deferred probe pending
```

samueldr commented:
Can you check for your issues against

 - the current stable release (2022.07-006)
 - stock U-Boot [2022.07, 2023.07]

Reporting a matrix of where things break and fail under as much of the same conditions would help figure out if it's a regression, and if it is, where it may come from.

shvetsnikita commented:
> Reporting a matrix of where things break and fail under as much of the same conditions would help figure out if it's a regression, and if it is, where it may come from.

Tow-Boot on Renegade Elite:
- SPI installer works on Tow-Boot 2022.07, but not on Tow-Boot 2023.07 - installer fail to perform writing image to SPI. 
- If I press Esc (on directly attached USB keyboard) when Tow-Boot 2022.07 propose to do so, it'll drop me to the menu. If I press Esc when Tow-Boot 2023.07 propose to do so, I either getting no response (Tow-Boot proceed with boot of installed OS from eMMC) or it'll drop me to the console, but never to the menu.
- If I boot Linux 6.1 (Debian stable) or 6.4 (Debian unstable) with Tow-Boot 2022.07 - USB works (albeit with [some issues](https://github.com/Tow-Boot/Tow-Boot/issues/209#issuecomment-1660388599), but overall it works) but if I boot same kernels with Tow-Boot 2023.07 - I getting [USB-related errors in dmesg](https://github.com/Tow-Boot/Tow-Boot/pull/255#issuecomment-1660391979) and empty lsusb. It's possible to workaround this specific issue with Tow-Boot 2023.07 by supplying kernel's own dtb from /lib/firmware via grub "devicetree" command - then USB works on Tow-Boot 2023.07 too.

Upstream U-Boot on Renegade Elite (I used [this project](https://github.com/johang/sd-card-images) to build img):
- With U-Boot 2022.07 USB doesn't work in U-Boot and GRUB2 (no even directly attached keyboard) but works in Linux.
- With U-Boot 2023.07 USB works in U-Boot and GRUB2, but not in Linux, just like with Tow-Boot 2023.07

So not working USB in Linux is seems like upstream regression. As for SPI installer and menu in Tow-Boot 2023.07 - I guess it's Tow-Boot own regressions?

samueldr commented:
Oh, the menu not working *may* be related to the move off of distroboot for rockchip hardware; the Tow-Boot menu is built using distroboot semantics, and this is the main thing we need to fix. So for any rk3399 targets, I'd say wait until that has been "finished" on my end.

And yes, if mainline U-Boot shows the same regression in equivalent builds, your assumption here is right: it's likely something caused within U-Boot.

samueldr committed:
tow-boot/builder: Drop EXTRAVERSION, always

It is assumed we follow the release, and the only SUBLEVEL release they
did was to fix the EXTRAVERSION being left from the last RC.

To remove complexity with making `2023.07.02` expose itself as
`2023.07`, we'll instead ignore the problem for now.

if/when they do a SUBLEVEL release that warrants exposing we are using
it, we'll re-evaluate how we handle that.

samueldr committed:
hardware/rockchip: Only built SPI image for systems with SPI

samueldr committed:
tow-boot/config: Ensure distroboot is used (for now)

We'll have to wait for a critical mass of devices we support to be
using bootstd before we flip the script and instead use bootstd
semantics as our source of truth.

samueldr committed:
ELO: Fix pinephoneA64 source ref

The `megous` user on github vanished. Let's re-use the same revision ID
from a known repo.

samueldr committed:
uBoot-sandbox: Ensure it builds

The test functions for bootstd require bootstd to be built in.

samueldr committed:
modules/tow-boot: Implement tagged releases

This is necessary since we want to keep track of changes between actual
releases, so e.g. have `-rcX` in `development` during the lifecycle of
the upgrade cycle, but being able to send fixes to the U-Boot codebase.

By using tags for every time we change the U-Boot codebase, we can more
easily `diff` between our changes.

samueldr committed:
support/release.sh: Ensure we don't have a stray `-rc`
